Make wxOverlay less dependent on external wxDCs

AliKet opened the pull request:
Working with `wxOverlay` requires three entities:
- a `wxOverlay` object (which hides port-specific code to create the actual overlay window)
- a `wxDC` to draw on
- a `wxDCOverlay` to connect the overlay with the drawing dc

With this PR the user is no longer required to specify the drawing dc explicitly.
i.e. if dc is omitted (see the new added ctors), a `wxClientDC` or `wxScreenDC` will be created internally for use. this will cover all the overlays used internally by the library itself.

The aim is to remove the dependency on the external dc and just use a `wxMemoryDC` instead.
AFAICT **wxOverlay** under `wxMSW` and `wxGTK` (and `wxMAC` ?) can simply use `wxMemoryDC` and blit the result.

If you agree to this, i'll close the other PRs ([this](https://github.com/wxWidgets/wxWidgets/pull/2448) and [this](https://github.com/wxWidgets/wxWidgets/pull/2386)) and continue with this one.

TIA.

Commit:
Stop requiring an external dc to create wxOverlay

With these changes, wxDCOverlay can be used just like a real dc
(thanks to the implicit conversion to wxDC) to draw into.

If the port-specific implementation supports the creation of wxOverlay
without any external dc (in which case wxOVERLAY_NO_EXTERNAL_DC must be defined)
wxDCOverlay creates a wxMemoryDC which get eventualy blitted onto the overlay window.

Commit:
Update drawing sample

Commit:
Make wxOverlay (generic implementation) available in all platforms

This can be useful to implement wxCaret (for ex.) where the generic
implementation is more efficient than the native one

AliKet commented:
@vadz @csomor 

If the changes here look good/correct to you then I want you to know that this commit is a breaking change because the user code should be updated like this for things to work correctly with the (generic) overlay:
- save the current (old) position of the caret
- move the caret to the new position
- draw the character at the old position

I tested this under wxGTK and it works correctly (Wayland needs a native implementation though). under wxMSW works too as before. but have no idea how this behaves under wxOSX as i don't have a Mac machine to test this myself.

So I would be grateful if you could test this for me (when you have time of course) under wxOSX and tell me how it goes there.

TIA.

vadz commented:
Just to be sure I understand correctly: when you say that that commit is a breaking change, what exactly do you mean, i.e. what will happen to the existing code using `wxOverlay` if these changes are merged? Will it stop compiling? Will it still compile but stop working? If it's the former, we're going to at least need some clear instructions in the docs for updating the code, but the latter is to be avoided at all costs. Of course, earlier you wrote that the existing code should continue to work unchanged, so maybe I'm just misunderstanding you?

Also, I think it would be wrong to add new API without support for Wayland or, at least, without a clear way to implement it under Wayland because we're clearly going to need it.

AliKet commented:
> Just to be sure I understand correctly: when you say that that commit is a breaking change, what exactly do you mean, i.e. what will happen to the existing code using `wxOverlay` if these changes are merged? Will it stop compiling? Will it still compile but stop working? If it's the former, we're going to at least need some clear instructions in the docs for updating the code, but the latter is to be avoided at all costs.

I mean the latter, unfortunately!

> Of course, earlier you wrote that the existing code should continue to work unchanged, so maybe I'm just misunderstanding you?
> 
Sorry, I wanted to clean up the old code and remove it entirely from `wxCaret` as it does exactly what the generic overlay does. Then I thought it was better to save what is under the caret just once and only reset it when needed.
i.e. while the caret is not moving and blinking in a given position, there is no need to save what is underneath it again and again needlessly (this is what the old code does currently, see `wxCaret::Refresh()`). instead, if we move the caret first and draw in its old position than:
- we can just assume that the caret always has a valid backup of what was in the new position before the caret moves.
- no need to suspend the caret before drawing at its position
- we save a lot of unneeded blitting

it's likely that I am overthinking it too much, so i'll revert it and do what the old code does.

> Also, I think it would be wrong to add new API without support for Wayland or, at least, without a clear way to implement it under Wayland because we're clearly going to need it.

AFAICT with [this](https://github.com/wxWidgets/wxWidgets/pull/2386) `wxCaret` works fine under Wayland (just needs to be updated with the changes made here). I'm still not sure if I should close that PR and continue over here, or keep things separate for easier review ?

Commit:
Update wxCaret to use generic wxOverlay

This has the benefit of being slightly more efficient
than using native overlay (if the generic version ever
works on the target platform. i.e. it doesn't work under
Wayland which needs a native wxOverlay anyhow)

Commit:
caret sample fixup

CreateCaret() should correctly handle caret re-creations

AliKet commented:
@vadz 

This commit a75f2a0 is literally the same compared to the (old) now removed code,
and shouldn't introduce any regression AFAICS.

**Note**: If the generic overlay doesn't work on wxOSX then we can guard this line:
    `m_overlay.UseGeneric();`
from being defined there.

TIA.

AliKet commented:
> @vadz
> 
> This commit [a75f2a0](https://github.com/wxWidgets/wxWidgets/commit/a75f2a06667f84b0b4fd3fd1d06562e3e1c514ba) is literally the same compared to the (old) now removed code,
> and shouldn't introduce any regression AFAICS.
> 
By "the (old) now removed code" I mean the original code.

Commit:
Ensure that wxCaret still works with native overlays

AliKet commented:
replaced by  #2522
